feat(compiler): add diagnostic for unused standalone imports

crisbeto opened the pull request:
Adds a new diagnostic that will report cases where a declaration is in the `imports` array, but isn't being used anywhere. The diagnostic is reported as a warning by default and can be controlled using the following option in the tsconfig:

```
{
  "angularCompilerOptions": {
    "extendedDiagnostics": {
      "checks": {
        "unusedStandaloneImports": "suppress"
      }
    }
  }
}
```

Fixes #46766.

crisbeto commented:
Pushed a change to add an automatic code fix to the language service.

crisbeto commented:
Caretaker note: these changes require a new patch internally. The patch is in [cl/670184686](http://cl/670184686)

atscott commented on adev/src/content/reference/extended-diagnostics/NG8113.md:
```diff
@@ -0,0 +1,48 @@
+# Unused Standalone Declarations
```
bikeshedding nit: I'm curious about the choice of "Declarations" instead of "Import(s)".

crisbeto replied:
Good point, I've renamed it everywhere.

crisbeto committed:
feat(compiler): add diagnostic for unused standalone imports

Adds a new diagnostic that will report cases where a declaration is in the `imports` array, but isn't being used anywhere. The diagnostic is reported as a warning by default and can be controlled using the following option in the tsconfig:

```
{
  "angularCompilerOptions": {
    "extendedDiagnostics": {
      "checks": {
        "unusedStandaloneImports": "suppress"
      }
    }
  }
}
```

**Note:** I'll look into a codefix for the language service in a follow-up.

Fixes #46766.

crisbeto committed:
feat(language-service): add code fix for unused standalone imports

Adds an automatic code fix to the language service that will remove unused standalone imports.

alxhub reviewed (approved):
Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
This PR was merged into the repository by commit 8da9fb49b54e50de2d028691f73fb773def62ecd.

The changes were merged into the following branches: main
